Dockerfile

```diff
@@ -1,11 +1,28 @@
-FROM node:16-alpine@sha256:72a490e7ed8aed68e16b8dc8f37b5bcc35c5b5c56ee3256effcdee63e2546f93 AS builder
+
+# An elaborated scheme to build all the dependencies of all packages first in a cached layer
+# https://stackoverflow.com/a/63142468/134409
+# https://medium.com/@emilefugulin/building-a-sane-docker-image-for-typescript-lerna-and-prisma-2-76d8ff9926e4
+FROM node:16-alpine@sha256:72a490e7ed8aed68e16b8dc8f37b5bcc35c5b5c56ee3256effcdee63e2546f93 AS filter-packages-json
 MAINTAINER Tyler Levine <tyler@bitgo.com>
+
+COPY package.json yarn.lock lerna.json ./
+WORKDIR /tmp/bitgo
+COPY package.json yarn.lock lerna.json ./
+COPY modules ./modules
+# delete all the non package.json files under `./modules/`
+RUN find modules \! -name "package.json" -mindepth 2 -maxdepth 2 -print | xargs rm -rf
+
+FROM node:16-alpine@sha256:72a490e7ed8aed68e16b8dc8f37b5bcc35c5b5c56ee3256effcdee63e2546f93 AS builder
 RUN apk add --no-cache git python3 make g++ libtool autoconf automake
 WORKDIR /tmp/bitgo
+COPY --from=filter-packages-json /tmp/bitgo .
+# (skip postinstall) https://github.com/yarnpkg/yarn/issues/4100#issuecomment-388944260
+RUN NOYARNPOSTINSTALL=1 yarn install --pure-lockfile
```

Comment:
wouldn't you want to do this in the `filter-packages-json` step? What's the point of copying that content in a separate layer?

Reply:
I'm like 99% sure, but basically that is the root of the trick: Docker caching is FS content based. In `filter-packages-json` I have to copy whole `./modules` which will invalidate `filter-packages-json` and thus re-run it very often. But this step is very fast, and deletes most of the files. So by having a new docker build image, I can copy only the filtered files, which usually will not change. Thus this layer will not be invalidated, even if `filter-packages-json` was. So everything here will be cached and not rebuild 99% of the time.